Added the loading of save data for any level

jmanplaysgames opened the pull request:
Hello again, github was being wierd and I couldnt commit any changes to my fork unless I deleted the repo and forked the newest version of Super-Haxagon.

I needed to make my old code it compatable with the new custom level format and add some things to your code. It now uses the name of the level for the save file, while also categorizing the save data in the following folder format to try to minimize the amount of conflicting save data : "Creator/Difficulty/Mode/LevelName.txt". There shouldnt be any conflicting save data issues unless somebody, for some reason, creates two different packs of levels that have the same name and difficulty as one they've done before.
They would need to get creative with the names or change the name of the creator to keep compatability with other things they could make. 

I could probably restructure the code a little bit, but Id like to get this in for version 2.0 as soon as i can since its so close to being completed. Everything appears to be working on my end, but please tell me if you have any problems with it.

jmanplaysgames commented:
Sorry It took so long, it was a mix of spending the holidays with my family and hardcore procrastination at its finest.

RedInquisitive commented:
Hmm, my implementation was to save all of the scores in one db file.  Your implementation is quite... complicated. I'll review it and see what implementation I want to use!

jmanplaysgames commented:
Yeah, i guess it might be, huh. I thought it would be easier this way because I wasn't sure how to just change 1 line of the file without clearing it first if it had multiple scores in it. As far as im aware, though I may be wrong, you cant remove text from a file, only add to it or clear the entire file. So if you had one giant file of scores, you would have to first copy it to a second file, while also making sure you replace the correct score.
It doesn't matter to me which one you use, choose whatever one you think is better.

RedInquisitive commented:
You are correct, you can't change one line of the file.  In order to update the score, you just  simply save all of the scores you have in memory (therefore you  only  ever need to read the file once). If you  really  wanted to make it efficient,  you could write to the  file in binary mode if you know the offset in the file, but  the speed at witch a 3DS can  write an ~500byte file makes that math worthless , so it's probably just  better to re-write the whole file.

...but right now I have to figure out why Citra is spitting out thousands of "unmapped Write32"s when I  launch a level ☹️

jmanplaysgames commented:
That was happening to me when it was constantly checking for the save file. I made it check for the save file only when you change which level is selected. Otherwise, depending on where you have it check for the save file, it might be checking for it every frame, which isnt very good.

RedInquisitive commented:
No, I don't think it's that, I actually think  it's hitting the stack limit  for some reason. I'm trying  to  attach a debugger to citra but that's harder  than it sounds...

jmanplaysgames commented:
Oh, i just reread you previous message and I read it wronf. I got that issue when I added a method call to the Main method. But only when I removed it, it started working again.

RedInquisitive commented:
It has something  to do with entering the  `doPlayGame(...)` method and I have no idea  what  is  triggering  it. I put an `exit(1)` before that method and Citra closes, but when I put `exit(1)` as the first line inside `doPlayGame(...)`  it hangs on the memory error! 

C makes no  sense  right now 😠

jmanplaysgames commented:
Yeah, I was just going off of my experience. When I called my method that checks to see if the save data is there inside of the main method, it caused the game to crash as soon as a level was loaded. It was strange since I thought it would run when the game was started, but the moment I removed that call the level loaded correctly. So if you added a call to something in the main method, try to find a different place you can put it that will still make it work.

RedInquisitive commented:
If  you want to hear the  real kicker, changing `#define TOTAL_PATTERNS_AT_ONE_TIME 3`  from 3 to 2 will fix the problem.

WHAT?

jmanplaysgames commented:
That's certainly a wierd fix, but I guess as long as it still works, right?

RedInquisitive commented:
I mean,  you'll only see two  patterns into the future instead of  three,  so patterns  will  randomly  appear (which is no good)! I'll see  what I can do.

RedInquisitive commented:
Welp I fixed it.... Turns  out this whole time it was `Track bgm`  that was being `audioFree()`'d before it was initialized,  causing Citra to freak out about uninitialized memory BUT ONLY if the stars aligned and the memory set by Citra was not already 0.

FML.

RedInquisitive commented:
Hey man, don't  take this the hard way, but I don't think I'm going  to implement scores like this. I was kind of focused on re-writing version 2.0 and at  the moment (as you  can  see) it's a little hard to accept new  pull requests. 

Truly, it means a lot to me that you contributed to the project (considering I thought it was dead to the outside world), but I'm working hard to make sure that version 2.0's base features are  implemented before I can accept any new ones. 

Please, please, please don't get discouraged from contributing to projects because of this! I'm 100% certain both of us learned more about C through this than if you never came along. I'll let everyone know when I am ready  to take new pull requests in the future (Most likely in a couple days,  no worries!)

Thanks for helping create a better  game for everyone!

jmanplaysgames commented:
Thats fine. As I said before, I dont mind which one you use. I just hope I helped out in some way.

RedInquisitive commented:
You've  helped out more than  enough! Seriously, like thanks  for everything!

Quick question, what made you stumble across this  repo? Most of  the posts I  made have been buried by new content haha (considering they  were made half a year ago).

jmanplaysgames commented:
We actually talked a little bit about 3DS hacking at the Prism Concert. I was the pleb that still had an emunand with A9lh while you had your CFW on your sysnand. 

One of my friends actually found your github account cause he wasn't sure how to do something for one of the Computer Science projects. Since he codes better with examples, he googled the project name and saw your repo. We both recognized your name, but later I checked out your account and saw you asked for help on this, so I tried to see if there was anything I could do to help.

So thats pretty much it. Unfortunately, I'm not some random person from the internet. But I do hope that other people play this.

RedInquisitive commented:
Aha, that makes more sense! Thanks for helping out and being a beta tester for my software 😄

jmanplaysgames commented:
No problem. I'll always be here, so I'll be keeping an eye on this project. To be honest, I didnt know anything about C before helping out with this. So I'm actually surprised that everything worked out the way it did, just by looking things up. So yeah, I'll always be here in case you need anything.

RedInquisitive commented:
I'm just surprised somebody other than me got this to compile without even asking me a single question! You just... did it!

Anyway, my implementation is [here](https://github.com/RedInquisitive/Super-Haxagon/blob/master/source/scores.c) if you want to see it!
